test/test_nn.py

```diff
@@ -6381,6 +6381,15 @@ def test_kl_div_with_diff_type(self):
                     result = torch.nn.functional.kl_div(input, target)
                     self.assertEqual(result.item(), expected.item(), 0.001)
 
+    def test_kl_div_log_softmax_target(self):
+        for device in device_():
+            a = torch.tensor([[1.0, 2, 3], [5.0, 5, 5]], device=device)
+            b = torch.tensor([[1.0, 2, 3], [5.0, 5, 5]], device=device)
+            self.assertEqual(
+                F.kl_div(F.log_softmax(a, 1), F.log_softmax(b, 1), reduction='none', log_target=True),
+                torch.zeros_like(a)
+            )
```

Comment:
This is not sufficient testing. You should at the very least also run gradcheck on the `log_target` version of `kl_div`. Grep for KLDiv in torch/testing/_internal/common_nn.py But I'd recommend reading through the existing KLDiv tests and seeing if there is anything else that seems like a good check for the log target case too.

Reply:
There are no tests for kl_div backward at all. Shall I add some?

Reply:
Also, even though the tests from `torch/testing/_internal/common_nn.py` kind of test the math, they are a bit meaningless, because KLDiv accepts only distributions, while one of the arguments is actually sampled from a Normal distribution. What about some additional checks for valid inputs in the `kl_div` maybe?

Reply:
They are "CriterionTest"s and yes they test gradients.

```
class NewCriterionTest(InputVariableMixin, CriterionTest):
    # TODO: check that criterions don't ignore grad_output
            
    def __init__(self, *args, **kwargs):
        super(NewCriterionTest, self).__init__(*args, **kwargs)
        self.check_gradgrad = kwargs.get('check_gradgrad', True)
        self.check_half = kwargs.get('check_half', True)
        self.check_bfloat16 = kwargs.get('check_bfloat16', False)
        self.convert_target = kwargs.get('convert_target', True)
                
    def _do_extra_tests(self, test_case, module, input, target):
        if not self.check_gradgrad:
            return
            
        test_case.assertFalse(target.requires_grad)
                
        params = tuple(x for x in module.parameters())
        if not isinstance(input, tuple):
            inputs = (input,) + params
                
            def apply_fn(input, *params):
                return module(input, target)
        else:
            inputs = input + params

            def apply_fn(input1, input2, *params):
                return module(input1, input2, target)
            
        # TODO: we don't pass `target` as part of inputs because we don't
        # currently compute the gradient w.r.t. target for loss functions.
        gradcheck(apply_fn, inputs)
        gradgradcheck(apply_fn, inputs)
```

Reply:
It's possible that the tests are meaningless, but I don't think where you sample the inputs from matter for a grad check, since you're just testing if analytical and numerical gradients match?

Reply:
But it is the formula for `target` that is affected... All right, I will add these tests separately. Thank you, now it makes much more sense to me :)

Reply:
> But it is the formula for target that is affected... 

Well, you are the expert! Sounds like CriterionTest will not actually exercise what you need it to exercise hahaha.